crates/types/src/consensus.rs

```diff
@@ -75,6 +78,17 @@ pub struct Consensus<TYPES: NodeType> {
 
     /// A reference to the metrics trait
     pub metrics: Arc<ConsensusMetricsValue>,
+
+    /// The most recent upgrade certificate this node formed.
+    /// Note: this is ONLY for certificates that have been formed internally,
+    /// so that we can propose with them.
+    ///
+    /// Certificates received from other nodes will get reattached regardless of this fields,
+    /// since they will be present in the leaf we propose off of.
+    pub formed_upgrade_certificate: Option<UpgradeCertificate<TYPES>>,
+
+    /// most recent decided upgrade certificate
+    pub decided_upgrade_cert: Option<UpgradeCertificate<TYPES>>,
```

Comment:
See my comment in lib.rs. It's probably better to have two `Consensus` structs to make sure we always pass in the correct fields.

If this is too complicated (shouldn't be?), I think we should at least add a comment here.

Reply:
This is just headaches on headaches. I added the separate struct, but initializing was not quite doable without defining a super `new` method or trying to implement Default so it would smooth over the fields that are default None. I'll add a comment and, for now, rename the variables to include something like `dontuse`